### PR TITLE
cudaPackages.cudnn: fixes libcublas not found error #236070

### LINKED ISSUE #236070: JaxlibWithCuda failed to load library libcublasLt.so.11
### Describe the bug
Jaxlib fails with the following message when Jax is trying to create new arrays.
```
Could not load library libcublasLt.so.11. Error: libcublasLt.so.11: cannot open shared object file: No such file or directory
Aborted (core dumped)
```

### Steps To Reproduce
1. use the following flake.nix to create an python environment with jax and jaxlib
```nix
{
  inputs = {
    nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.05";
  };

  outputs = { self, nixpkgs }:
    let
      system = "x86_64-linux";
      pkgs = import nixpkgs {
        inherit system;
        config = {
          allowUnfree = true;
          cudaSupport = true;
        };
      };
      python-env = pkgs.python310.withPackages (ps: with ps; [
        jax
        jaxlibWithCuda
      ]);
    in
    {
      devShells.${system}.default = pkgs.mkShell {
        buildInputs = [
          python-env
        ];
      };
    };
}
```
2. enter this environment
3. import jax and create an array:
`python -c "import jax.numpy as jnp; jnp.ones(10)"`

### Expected behavior
Jax should be able to create arrays on GPU.

### Additional context
I tried with Cuda version 11 and 12, both resulted in the same error.
And pytorch doesn't have this problem, so I think this is error is specific to jaxlib.

### Notify maintainers

@ndl 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.3.1-zen1, NixOS, 23.05 (Stoat), 23.05.20230604.e7603eb`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - channels(bill): `"home-manager, nixpkgs"`
 - nixpkgs: `/home/bill/.nix-defexpr/channels/nixpkgs`
```

### PR DESCRIPTION
**Pull request by ogoid**
###### Description of changes

Fixes #236070.

Based on [@BillHuang2001's debug](https://github.com/NixOS/nixpkgs/issues/236070#issuecomment-1586469070), `libcudnn_ops_infer.so` dloads `libcublas.so.11` and `libcublasLt.so.11`.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by BillHuang2001** [comment]
It seems patching `libcudnn_ops_infer.so` is not enough, there are other files depending on `libcublas.so.11`, `libcublasLt.so.11`
I did
```
patchelf $out/lib/*.so --add-needed libcublas.so.11
patchelf $out/lib/*.so --add-needed libcublasLt.so.11
```
which might be an overkill, but JAX won't run without it.

**Comment by ogoid** [comment]
For some reason just patching libcudnn_ops_infer.so is enough for me, but I added the suggested modification.

I also see a `libcublas` parameter in https://github.com/NixOS/nixpkgs/blob/aadb3f3058f8a93c5cc2b5de843d700d106218c1/pkgs/development/libraries/science/math/cudnn/generic.nix#L9.

Should we check if cublas is available somehow? What would be the proper way?

File: pkgs/development/libraries/science/math/cudnn/generic.nix [hunk 1]
```diff
@@ -94,6 +94,8 @@ in
     # Without --add-needed autoPatchelf forgets $ORIGIN on cuda>=8.0.5.
     postFixup = strings.optionalString (strings.versionAtLeast versionTriple "8.0.5") ''
       patchelf $out/lib/libcudnn.so --add-needed libcudnn_cnn_infer.so
+      patchelf $out/lib/*.so --add-needed libcublas.so.11
+      patchelf $out/lib/*.so --add-needed libcublasLt.so.11
```
Comment (samuela) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:98]:
```suggestion
    patchelf $out/lib/libcudnn_ops_infer.so --add-needed libcublas.so.11 --add-needed libcublasLt.so.11
```
As some others suggested, I found that this is all that's necessary to resolve the bug. Patching everything in `$out/lib` is a bit heavy handed imho.

**Review by samuela — CHANGES_REQUESTED** [review]
Thanks for putting this together @ogoid!

**Comment by ogoid** [comment]
> Thanks for putting this together @ogoid!

You should thank @BillHuang2001, as he's the one who debugged it.

**Comment by samuela** [comment]
Thanks @BillHuang2001! 🎉

File: pkgs/development/libraries/science/math/cudnn/generic.nix [hunk 2]
```diff
@@ -94,6 +94,7 @@ in
     # Without --add-needed autoPatchelf forgets $ORIGIN on cuda>=8.0.5.
     postFixup = strings.optionalString (strings.versionAtLeast versionTriple "8.0.5") ''
       patchelf $out/lib/libcudnn.so --add-needed libcudnn_cnn_infer.so
+    patchelf $out/lib/libcudnn_ops_infer.so --add-needed libcublas.so.11 --add-needed libcublasLt.so.11
```
Comment (samuela) [inline pkgs/development/libraries/science/math/cudnn/generic.nix:97]: Ahh I may have spoke too soon in making this suggestion... I fear that this won't work with cuda versions != 11... we should test it

Comment (samuela) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:97]: I wonder if
```
patchelf $out/lib/libcudnn_ops_infer.so --add-needed libcublas.so --add-needed libcublasLt.so
```
would work following the style of the preceding line

Comment (ogoid) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:97]: Also, is cublas always available? Doesn't seem so from here:

https://github.com/NixOS/nixpkgs/blob/e1c91b230122dceca76e600aae5a97d2f590ebb2/pkgs/development/libraries/science/math/cudnn/generic.nix#L9

Comment (samuela) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:97]: I'm not too worried about that... IIUC cublas comes included in cudatoolkit for versions <=10. but we should just test it on cudaPackages_10 and find out

Comment (samuela) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:97]: attempted to test on `cudaPackages_10` but ran into https://github.com/NixOS/nixpkgs/issues/245218

Comment (samuela) [reply pkgs/development/libraries/science/math/cudnn/generic.nix:97]: Ok just tested on `cudaPackages_12` and indeed it fails. However, this fix works on both 11 and 12:
```suggestion
      patchelf $out/lib/libcudnn_ops_infer.so --add-needed libcublas.so --add-needed libcublasLt.so
```

**Review by samuela — APPROVED** [review]
LGTM

cc @NixOS/cuda-maintainers

**Comment by samuela** [comment]
I'll go ahead and squash/merge tomorrow unless anyone objects?

**Review by ConnorBaker — APPROVED** [review]

**Pull request merged**